library/core/src/main/java/com/google/android/exoplayer2/scheduler/Requirements.java

```diff
@@ -162,6 +186,54 @@ private boolean isDeviceIdle(Context context) {
         : Util.SDK_INT >= 20 ? !powerManager.isInteractive() : !powerManager.isScreenOn();
   }
 
+  /**
+   * Implementation taken from the the WorkManager source.
+   * @see <a href="https://android.googlesource.com/platform/frameworks/support/+/androidx-master-dev/work/workmanager/src/main/java/androidx/work/impl/constraints/trackers/BatteryNotLowTracker.java">BatteryNotLowTracker</a>
+   */
+  private boolean isBatteryNotLow(Context context) {
```

Comment:
If it happens infrequently, it may be acceptable to use `ACTION_BATTERY_CHANGED` in `RequirementsWatcher` to get them aligned. This presumably also solves the problem below about the plugged state, since I'm guessing that *is* fired when the plugged state changes?

Reply:
I will retest the battery part more thoroughly, I've been testing on emulators and the Intents fired consistently between `ACTION_BATTERY_LOW` and `ACTION_BATTERY_OK`. What I did not test is a device where the battery low percentage is not 15% (my personal device starts warning at 20% for example). Good idea to look into the frequency of `ACTION_BATTERY_CHANGED` as well, will get back to you on the battery part.

I'd prefer to submit the storage and battery changes, including refactor, as one change if that's OK with you.

Reply:
One change sounds fine provided you can get the battery part to a state where it looks robust, without the need to additional refactoring. If additional refactoring is needed then we should start splitting it into smaller changes, which are easier to review and merge.

Reply:
Did some more thorough testing on the battery part and have the following findings:

>  Is it possible to use ACTION_BATTERY_LOW and ACTION_BATTERY_OK for the evaluation, rather than ACTION_BATTERY_CHANGED? 

This is indeed not possible as `ACTION_BATTERY_LOW` and `ACTION_BATTERY_OK` don't contain the intent extras needed for the evaluation.

> If it happens infrequently, it may be acceptable to use ACTION_BATTERY_CHANGED in RequirementsWatcher to get them aligned

`ACTION_BATTERY_CHANGED` fires on every "tick" of the battery level, as well as on connect and disconnect from a power source as you expected. The intent does fire more often but it is a more robust solution as `ACTION_BATTERY_LOW` and `ACTION_BATTERY_OK` seem to be device specific with respect to when they fire. For example on another test device the battery is only considered low when there's 10% left. Additionally I've seen some instances where `ACTION_BATTERY_OK` doesn't fire immediately after the battery level goes above the low threshold. If we use `ACTION_BATTERY_CHANGED` the behaviour will be consistent across devices using the same magic value we choose as the low threshold.